File: src/test/java/io/vproxy/vfx/intro/IntroFXMain.java
package io.vproxy.vfx.intro;

import io.vproxy.vfx.control.globalscreen.GlobalScreenUtils;
import io.vproxy.vfx.manager.image.ImageManager;
import io.vproxy.vfx.manager.task.TaskManager;
import io.vproxy.vfx.theme.Theme;
import io.vproxy.vfx.ui.button.FusionButton;
import io.vproxy.vfx.ui.button.FusionImageButton;
import io.vproxy.vfx.ui.layout.HPadding;
import io.vproxy.vfx.ui.layout.VPadding;
import io.vproxy.vfx.ui.pane.FusionPane;
import io.vproxy.vfx.ui.scene.*;
import io.vproxy.vfx.ui.stage.VStage;
import io.vproxy.vfx.util.FXUtils;
import javafx.application.Application;
import javafx.geometry.Insets;
import javafx.scene.layout.*;
import javafx.stage.Stage;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class IntroFXMain extends Application {
    private final List<DemoVScene> mainScenes = new ArrayList<>();
    private VSceneGroup sceneGroup;

    @Override
    public void start(Stage primaryStage) {
        ImageManager.get().loadBlackAndChangeColor("images/menu.png", Map.of("white", 0xffffffff));
        ImageManager.get().loadBlackAndChangeColor("images/up-arrow.png", Map.of("white", 0xffffffff));

        var stage = new VStage(primaryStage) {
            @Override
            public void close() {
                super.close();
                TaskManager.get().terminate();
                GlobalScreenUtils.unregister();
            }
        };
        stage.getInitialScene().enableAutoContentWidthHeight();

        stage.setTitle("VFX Intro");

        mainScenes.add(new _00IntroScene());
        mainScenes.add(new _01aVStageIntroScene());
        mainScenes.add(new _01bVStageInitParamsScene());
        mainScenes.add(new _01cVStageStructureScene());
        mainScenes.add(new _02aVSceneGroupIntroScene());
        mainScenes.add(new _02bVSceneGroupDisplayScene());
        mainScenes.add(new _02cVSceneGroupDemoScene(() -> sceneGroup));
        mainScenes.add(new _02dVSceneGroupMenuScene());
        mainScenes.add(new _02eVSceneGroupStructureScene());
        mainScenes.add(new _03aVSceneIntroScene());
        mainScenes.add(new _03bVSceneStructureScene());
        mainScenes.add(new _04aVScrollPaneIntroScene());
        mainScenes.add(new _04bVScrollPaneDemoScene());
        mainScenes.add(new _04cVScrollPaneStructureScene());
        mainScenes.add(new _05aFusionPaneIntroScene());
        mainScenes.add(new _05bFusionPaneDemoScene());
        mainScenes.add(new _05cFusionPaneStructureScene());
        mainScenes.add(new _06aVTableViewIntroScene());
        mainScenes.add(new _06bVTableViewDemoScene());
        mainScenes.add(new _06cVTableViewDemo2Scene());
        mainScenes.add(new _06dVTableViewStructureScene());
        mainScenes.add(new _06eVTableViewSpecialUsageScene());
        mainScenes.add(new _07aAnimationSystemIntroScene());
        mainScenes.add(new _07bAnimationSystemDemoScene());
        mainScenes.add(new _07cAnimationSystemDescScene());
        mainScenes.add(new _08aLoadingStageIntroScene());
        mainScenes.add(new _08bLoadingStageDemoScene());
        mainScenes.add(new _08cLoadingStageStructureScene());
        mainScenes.add(new _xxaComponentsIntroScene());
        mainScenes.add(new _xxbComponentsDemoScene());
        mainScenes.add(new _xxcComponentsDemo2Scene());
        mainScenes.add(new _xxdComponentsDemo3Scene());
        mainScenes.add(new _zzzzzEndingScene());

        // var initialScene = mainScenes.stream().filter(e -> e instanceof ).findAny().get();
        var initialScene = mainScenes.get(0);
        sceneGroup = new VSceneGroup(initialScene);
        for (var s : mainScenes) {
            if (s == initialScene) continue;
            sceneGroup.addScene(s);
        }

        var navigatePane = new FusionPane();

        navigatePane.getNode().setPrefHeight(60);
        FXUtils.observeHeight(stage.getInitialScene().getContentPane(), sceneGroup.getNode(), -10 - 60 - 5 - 10);

        FXUtils.observeWidth(stage.getInitialScene().getContentPane(), sceneGroup.getNode(), -20);
        FXUtils.observeWidth(stage.getInitialScene().getContentPane(), navigatePane.getNode(), -20);

        var prevButton = new FusionButton("<< Previous") {{
            setPrefWidth(150);
            setPrefHeight(navigatePane.getNode().getPrefHeight() - FusionPane.PADDING_V * 2);
            setOnlyAnimateWhenNotClicked(true);

            var current = sceneGroup.getCurrentMainScene();
            //noinspection SuspiciousMethodCalls
            var index = mainScenes.indexOf(current);
            if (index == 0) {
                setDisable(true);
            }
        }};
        var nextButton = new FusionButton("Next >>") {{
            setPrefWidth(150);
            setPrefHeight(navigatePane.getNode().getPrefHeight() - FusionPane.PADDING_V * 2);
            setOnlyAnimateWhenNotClicked(true);

            var current = sceneGroup.getCurrentMainScene();
            //noinspection SuspiciousMethodCalls
            var index = mainScenes.indexOf(current);
            if (index == mainScenes.size() - 1) {
                setDisable(true);
            }
        }};
        prevButton.setOnAction(e -> {
            var current = sceneGroup.getCurrentMainScene();
            //noinspection SuspiciousMethodCalls
            var index = mainScenes.indexOf(current);
            if (index == 0) return;
            sceneGroup.show(mainScenes.get(index - 1), VSceneShowMethod.FROM_LEFT);
            if (index - 1 == 0) {
                prevButton.setDisable(true);
            }
            nextButton.setDisable(false);
        });
        nextButton.setOnAction(e -> {
            var current = sceneGroup.getCurrentMainScene();
            //noinspection SuspiciousMethodCalls
            var index = mainScenes.indexOf(current);
            if (index == mainScenes.size() - 1) return;
            sceneGroup.show(mainScenes.get(index + 1), VSceneShowMethod.FROM_RIGHT);
            if (index + 1 == mainScenes.size() - 1) {
                nextButton.setDisable(true);
            }
            prevButton.setDisable(false);
        });

        navigatePane.getContentPane().getChildren().add(prevButton);
        navigatePane.getContentPane().getChildren().add(nextButton);
        navigatePane.getContentPane().widthProperty().addListener((ob, old, now) -> {
            if (now == null) return;
            var v = now.doubleValue();
            nextButton.setLayoutX(v - nextButton.getPrefWidth());
        });

        var box = new HBox(
            new HPadding(10),
            new VBox(
                new VPadding(10),
                sceneGroup.getNode(),
                new VPadding(5),
                navigatePane.getNode()
            )
        );
        stage.getInitialScene().getContentPane().getChildren().add(box);

        var menuScene = new VScene(VSceneRole.DRAWER_VERTICAL);
        menuScene.getNode().setPrefWidth(450);
        menuScene.enableAutoContentWidth();
        menuScene.getNode().setBackground(new Background(new BackgroundFill(
            Theme.current().subSceneBackgroundColor(),
            CornerRadii.EMPTY,
            Insets.EMPTY
        )));
        stage.getRootSceneGroup().addScene(menuScene, VSceneHideMethod.TO_LEFT);
        var menuVBox = new VBox() {{
            setPadding(new Insets(0, 0, 0, 24));
            getChildren().add(new VPadding(20));
        }};
        menuScene.getContentPane().getChildren().add(menuVBox);
        for (int i = 0; i < mainScenes.size(); ++i) {
            final var fi = i;
            var s = mainScenes.get(i);
            var title = s.title();
            var button = new FusionButton(title);
            button.setDisableAnimation(true);
            button.setOnAction(e -> {
                //noinspection SuspiciousMethodCalls
                var currentIndex = mainScenes.indexOf(sceneGroup.getCurrentMainScene());
                if (currentIndex != fi) {
                    sceneGroup.show(s, currentIndex < fi ? VSceneShowMethod.FROM_RIGHT : VSceneShowMethod.FROM_LEFT);
                }
                stage.getRootSceneGroup().hide(menuScene, VSceneHideMethod.TO_LEFT);
                prevButton.setDisable(fi == 0);
                nextButton.setDisable(fi == mainScenes.size() - 1);
            });
            button.setPrefWidth(400);
            button.setPrefHeight(40);
            if (i != 0) {
                menuVBox.getChildren().add(new VPadding(20));
            }
            menuVBox.getChildren().add(button);
        }
        menuVBox.getChildren().add(new VPadding(20));

        var menuBtn = new FusionImageButton(ImageManager.get().load("images/menu.png:white")) {{
            setPrefWidth(40);
            setPrefHeight(VStage.TITLE_BAR_HEIGHT + 1);
            getImageView().setFitHeight(15);
            setLayoutX(-2);
            setLayoutY(-1);
        }};
        menuBtn.setOnAction(e -> stage.getRootSceneGroup().show(menuScene, VSceneShowMethod.FROM_LEFT));
        stage.getRoot().getContentPane().getChildren().add(menuBtn);

        stage.getStage().setWidth(1280);
        stage.getStage().setHeight(800);
        stage.getStage().centerOnScreen();
        stage.getStage().show();
    }
}

File: src-ts/src/intro/IntroFXMain.ts
import { ImageManager } from '../manager/image/ImageManager.js';
import { Theme } from '../theme/Theme.js';
import { FusionButton } from '../ui/button/FusionButton.js';
import { FusionImageButton } from '../ui/button/FusionImageButton.js';
import { HPadding } from '../ui/layout/HPadding.js';
import { VPadding } from '../ui/layout/VPadding.js';
import { FusionPane } from '../ui/pane/FusionPane.js';
import { VScene } from '../ui/scene/VScene.js';
import { VSceneGroup } from '../ui/scene/VSceneGroup.js';
import { VSceneRole } from '../ui/scene/VSceneRole.js';
import { VSceneShowMethod } from '../ui/scene/VSceneShowMethod.js';
import { VSceneHideMethod } from '../ui/scene/VSceneHideMethod.js';
import { VStage } from '../ui/stage/VStage.js';
import { FXUtils } from '../util/FXUtils.js';
import { Parent } from '../javafx/Parent.js';
import { HBox } from '../javafx/HBox.js';
import { VBox } from '../javafx/VBox.js';
import {
  Background,
  BackgroundFill,
  CornerRadii,
  Insets,
} from '../javafx/layout.js';
import type { DemoVScene } from './DemoVScene.js';
import {
  Intro00Scene,
  VStage01aIntroScene,
  VStage01cStructureScene,
  VSceneGroup02aIntroScene,
  VSceneGroup02eStructureScene,
  VScene03aIntroScene,
  VScene03bStructureScene,
  VScrollPane04aIntroScene,
  VScrollPane04cStructureScene,
  FusionPane05aIntroScene,
  FusionPane05cStructureScene,
  VTableView06aIntroScene,
  VTableView06dStructureScene,
  VTableView06eSpecialUsageScene,
  Animation07aIntroScene,
  Animation07cDescScene,
  Loading08aIntroScene,
  Loading08cStructureScene,
  LogConsole09aIntroScene,
  ComponentsXxaIntroScene,
  EndingScene,
} from './scenesText.js';
import {
  VStage01bInitParamsScene,
  VSceneGroup02bDisplayScene,
  VSceneGroup02cDemoScene,
  VSceneGroup02dMenuScene,
  VScrollPane04bDemoScene,
  FusionPane05bDemoScene,
  VTableView06bDemoScene,
  VTableView06cDemo2Scene,
  Loading08bDemoScene,
  LogConsole09bDemoScene,
  ComponentsXxbDemoScene,
  ComponentsXxcDemo2Scene,
  ComponentsXxdBrokenLineScene,
  Animation07bDemoScene,
} from './scenesDemo.js';

export class IntroFXMain {
  private readonly mainScenes: DemoVScene[] = [];
  private sceneGroup!: VSceneGroup;
  private stage!: VStage;

  async start(): Promise<void> {
    // Recolor the intro images (black → white) so the menu / arrow icons
    // are visible on the dark theme background.
    ImageManager.get().loadBlackAndChangeColor('intro/res/menu.png', { white: 0xffffffff });
    ImageManager.get().loadBlackAndChangeColor('intro/res/up-arrow.png', { white: 0xffffffff });

    this.stage = new VStage();
    const origClose = this.stage.close.bind(this.stage);
    this.stage.close = async (): Promise<void> => {
      await origClose();
    };
    this.stage.getInitialScene().enableAutoContentWidthHeight();

    this.stage.setTitle('VFX Intro');

    this.mainScenes.push(new Intro00Scene());
    this.mainScenes.push(new VStage01aIntroScene());
    this.mainScenes.push(new VStage01bInitParamsScene());
    this.mainScenes.push(new VStage01cStructureScene());
    this.mainScenes.push(new VSceneGroup02aIntroScene());
    this.mainScenes.push(new VSceneGroup02bDisplayScene());
    this.mainScenes.push(new VSceneGroup02cDemoScene(() => this.sceneGroup));
    this.mainScenes.push(new VSceneGroup02dMenuScene());
    this.mainScenes.push(new VSceneGroup02eStructureScene());
    this.mainScenes.push(new VScene03aIntroScene());
    this.mainScenes.push(new VScene03bStructureScene());
    this.mainScenes.push(new VScrollPane04aIntroScene());
    this.mainScenes.push(new VScrollPane04bDemoScene());
    this.mainScenes.push(new VScrollPane04cStructureScene());
    this.mainScenes.push(new FusionPane05aIntroScene());
    this.mainScenes.push(new FusionPane05bDemoScene());
    this.mainScenes.push(new FusionPane05cStructureScene());
    this.mainScenes.push(new VTableView06aIntroScene());
    this.mainScenes.push(new VTableView06bDemoScene());
    this.mainScenes.push(new VTableView06cDemo2Scene());
    this.mainScenes.push(new VTableView06dStructureScene());
    this.mainScenes.push(new VTableView06eSpecialUsageScene());
    this.mainScenes.push(new Animation07aIntroScene());
    this.mainScenes.push(new Animation07bDemoScene());
    this.mainScenes.push(new Animation07cDescScene());
    this.mainScenes.push(new Loading08aIntroScene());
    this.mainScenes.push(new Loading08bDemoScene());
    this.mainScenes.push(new Loading08cStructureScene());
    this.mainScenes.push(new LogConsole09aIntroScene());
    this.mainScenes.push(new LogConsole09bDemoScene());
    this.mainScenes.push(new ComponentsXxaIntroScene());
    this.mainScenes.push(new ComponentsXxbDemoScene());
    this.mainScenes.push(new ComponentsXxcDemo2Scene());
    this.mainScenes.push(new ComponentsXxdBrokenLineScene());
    this.mainScenes.push(new EndingScene());

    const initialScene = this.mainScenes[0]!;
    this.sceneGroup = new VSceneGroup(initialScene);
    for (const s of this.mainScenes) {
      if (s === initialScene) continue;
      this.sceneGroup.addScene(s);
    }

    const navigatePane = new FusionPane();
    navigatePane.getNode().setPrefHeight(60);

    FXUtils.observeHeight(this.stage.getInitialScene().getContentPane(), this.sceneGroup.getNode(), -10 - 60 - 5 - 10);
    FXUtils.observeWidth(this.stage.getInitialScene().getContentPane(), this.sceneGroup.getNode(), -20);
    FXUtils.observeWidth(this.stage.getInitialScene().getContentPane(), navigatePane.getNode(), -20);

    const prevButton = new FusionButton('<< Previous');
    prevButton.setPrefWidth(150);
    prevButton.setPrefHeight(navigatePane.getNode().getPrefHeight() - FusionPane.PADDING_V * 2);
    prevButton.setOnlyAnimateWhenNotClicked(true);
    {
      const current = this.sceneGroup.getCurrentMainScene();
      const index = this.mainScenes.indexOf(current as DemoVScene);
      if (index === 0) prevButton.setDisabled(true);
    }

    const nextButton = new FusionButton('Next >>');
    nextButton.setPrefWidth(150);
    nextButton.setPrefHeight(navigatePane.getNode().getPrefHeight() - FusionPane.PADDING_V * 2);
    nextButton.setOnlyAnimateWhenNotClicked(true);
    {
      const current = this.sceneGroup.getCurrentMainScene();
      const index = this.mainScenes.indexOf(current as DemoVScene);
      if (index === this.mainScenes.length - 1) nextButton.setDisabled(true);
    }

    prevButton.setOnAction(() => {
      const current = this.sceneGroup.getCurrentMainScene();
      const index = this.mainScenes.indexOf(current as DemoVScene);
      if (index === 0) return;
      this.sceneGroup.show(this.mainScenes[index - 1]!, VSceneShowMethod.FROM_LEFT);
      if (index - 1 === 0) prevButton.setDisabled(true);
      nextButton.setDisabled(false);
    });
    nextButton.setOnAction(() => {
      const current = this.sceneGroup.getCurrentMainScene();
      const index = this.mainScenes.indexOf(current as DemoVScene);
      if (index === this.mainScenes.length - 1) return;
      this.sceneGroup.show(this.mainScenes[index + 1]!, VSceneShowMethod.FROM_RIGHT);
      if (index + 1 === this.mainScenes.length - 1) nextButton.setDisabled(true);
      prevButton.setDisabled(false);
    });

    navigatePane.getContentPane().getChildren().add(prevButton);
    navigatePane.getContentPane().getChildren().add(nextButton);
    navigatePane.getContentPane().widthProperty.addListener((_o, now) => {
      if (now === null) return;
      nextButton.setLayoutX(now - nextButton.getPrefWidth());
    });

    const box = new HBox(
      new HPadding(10),
      new VBox(
        new VPadding(10),
        this.sceneGroup.getNode(),
        new VPadding(5),
        navigatePane.getNode(),
      ),
    );
    this.stage.getInitialScene().getContentPane().getChildren().add(box);

    const menuScene = new VScene(VSceneRole.DRAWER_VERTICAL);
    menuScene.getNode().setPrefWidth(450);
    menuScene.enableAutoContentWidth();
    menuScene.getNode().setBackground(new Background(new BackgroundFill(
      Theme.current().subSceneBackgroundColor(),
      CornerRadii.EMPTY,
      Insets.EMPTY,
    )));
    this.stage.getRootSceneGroup().addScene(menuScene, VSceneHideMethod.TO_LEFT);

    const menuVBox = new VBox(new VPadding(20));
    menuVBox.el.style.paddingLeft = '24px';
    menuScene.getContentPane().getChildren().add(menuVBox);

    for (let i = 0; i < this.mainScenes.length; i++) {
      const fi = i;
      const s = this.mainScenes[i]!;
      const title = s.title();
      const button = new FusionButton(title);
      button.setDisableAnimation(true);
      button.setOnAction(() => {
        const currentIndex = this.mainScenes.indexOf(this.sceneGroup.getCurrentMainScene() as DemoVScene);
        if (currentIndex !== fi) {
          this.sceneGroup.show(
            s,
            currentIndex < fi ? VSceneShowMethod.FROM_RIGHT : VSceneShowMethod.FROM_LEFT,
          );
        }
        this.stage.getRootSceneGroup().hide(menuScene, VSceneHideMethod.TO_LEFT);
        prevButton.setDisabled(fi === 0);
        nextButton.setDisabled(fi === this.mainScenes.length - 1);
      });
      button.setPrefWidth(400);
      button.setPrefHeight(40);
      if (i !== 0) menuVBox.getChildren().add(new VPadding(20));
      menuVBox.getChildren().add(button);
    }
    menuVBox.getChildren().add(new VPadding(20));

    const menuBtn = new FusionImageButton(ImageManager.get().load('intro/res/menu.png:white'));
    menuBtn.setPrefWidth(40);
    menuBtn.setPrefHeight(VStage.TITLE_BAR_HEIGHT + 1);
    menuBtn.getImageView().setFitHeight(15);
    menuBtn.setLayoutX(-2);
    menuBtn.setLayoutY(-1);
    menuBtn.setOnAction(() => {
      this.stage.getRootSceneGroup().show(menuScene, VSceneShowMethod.FROM_LEFT);
    });
    this.stage.getRoot().getContentPane().getChildren().add(menuBtn);

    await this.stage.getStage().setWidth(1280);
    await this.stage.getStage().setHeight(800);
    await this.stage.getStage().centerOnScreen();
    await this.stage.show();
  }
}
